src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+import java.util.List;
+
+/**
+ * KDF implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC
+ * 5869.  This implementation provides the complete Extract-then-Expand HKDF
+ * function as well as Extract-only and Expand-only variants.
+ */
+abstract class HkdfKeyDerivation extends KDFSpi {
+
+    protected final int hmacLen;
+    protected final String hmacAlgName;
+
+    /**
+     * The sole constructor.
+     *
+     * @param kdfParameters
+     *     the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    HkdfKeyDerivation(String hmacAlgName, int hmacLen,
+                      KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {
+        super(kdfParameters);
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(
+                "RFC 5869 has no parameters for its KDF algorithms");
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived, or specifies
+     *     a type of output that is not a key (e.g. raw data)
+     * @throws IllegalArgumentException
+     *     if {@code alg} is {@code null} or empty
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+
+        if (alg == null || alg.isEmpty()) {
+            throw new IllegalArgumentException(
+                "the algorithm for the resultant SecretKey may not be null or"
+                + " empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(derivationParameterSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived key material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        SecretKey inputKeyMaterial;
+        SecretKey salt;
```

Comment:
Looking at the implementation, it seems you can just use byte[] for `inputKeyMaterial` and `salt`. Why bother packaging the bytes into a `SecretKey` object and later calling `getEncoded()` to retrieve it again?

Reply:
Well, you can't handle this case and throws InvalidKeyException when there are such keys. I've tried making the "SecretKey" to "byte[]" change and it simplifies the code.